src/vm-vcpu-ref/src/x86_64/msrs.rs

```diff
@@ -0,0 +1,80 @@
+// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+use kvm_bindings::{kvm_msr_entry, Msrs};
+
+use crate::x86_64::msr_index::{
+    MSR_CSTAR, MSR_IA32_MISC_ENABLE, MSR_IA32_MISC_ENABLE_FAST_STRING, MSR_IA32_SYSENTER_CS,
+    MSR_IA32_SYSENTER_EIP, MSR_IA32_SYSENTER_ESP, MSR_IA32_TSC, MSR_KERNEL_GS_BASE, MSR_LSTAR,
+    MSR_STAR, MSR_SYSCALL_MASK,
+};
+
+/// Errors associated with operations on MSRs.
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    /// Failed to initialize MSRS.
+    CreateMsrs,
+}
+/// Specialized result type for operations on MSRs.
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// Creates and populates required MSR entries for booting Linux on X86_64.
+///
+/// # Example - Set boot MSRs
+///
+/// ```rust
+/// use kvm_ioctls::Kvm;
+/// use vm_vcpu_ref::x86_64::msrs::create_boot_msr_entries;
+///
+/// let kvm = Kvm::new().unwrap();
+/// let vm = kvm.create_vm().unwrap();
+/// let vcpu = vm.create_vcpu(0).unwrap();
+///
+/// vcpu.set_msrs(&create_boot_msr_entries().unwrap()).unwrap();
+/// ```
+pub fn create_boot_msr_entries() -> Result<Msrs> {
+    let msr_entry_default = |msr| kvm_msr_entry {
+        index: msr,
+        data: 0x0,
+        ..Default::default()
+    };
+
+    let raw_msrs = vec![
+        msr_entry_default(MSR_IA32_SYSENTER_CS),
+        msr_entry_default(MSR_IA32_SYSENTER_ESP),
+        msr_entry_default(MSR_IA32_SYSENTER_EIP),
+        // x86_64 specific msrs, we only run on x86_64 not x86.
+        msr_entry_default(MSR_STAR),
+        msr_entry_default(MSR_CSTAR),
+        msr_entry_default(MSR_KERNEL_GS_BASE),
+        msr_entry_default(MSR_SYSCALL_MASK),
+        msr_entry_default(MSR_LSTAR),
+        // end of x86_64 specific code
+        msr_entry_default(MSR_IA32_TSC),
+        kvm_msr_entry {
+            index: MSR_IA32_MISC_ENABLE,
+            data: u64::from(MSR_IA32_MISC_ENABLE_FAST_STRING),
+            ..Default::default()
+        },
+    ];
+
+    Msrs::from_entries(&raw_msrs).map_err(|_| Error::CreateMsrs)
```

Comment:
I wonder if there's any useful error information that we can propagate back to the caller instead of discarding it.

Reply:
The reason I left it without the inner error is because we would need to take an explicit dependency on vmm-sys-util (right now we have a transient dependency on vmm-sys-util through the fam-wrappers feature of kvm-bindings). The `Msrs::from_entries` can only fail due to capacity issues, and I was thinking that's unlikely to change in the future. I don't mind adding it either.